pkg/server/plugin/datastore/sql/migration.go

```diff
@@ -35,25 +65,81 @@ func migrateDB(db *gorm.DB, dbType string, log hclog.Logger) (err error) {
 	if err := db.Assign(Migration{}).FirstOrCreate(migration).Error; err != nil {
 		return sqlError.Wrap(err)
 	}
-	version := migration.Version
 
-	if version > codeVersion {
-		err = sqlError.New("backwards migration not supported! (current=%d, code=%d)", version, codeVersion)
-		log.Error(err.Error())
-		return err
+	schemaVersion := migration.Version
+
+	var dbCodeVersion semver.Version
+	// Auto-migrate can only be safely disabled if 0.9.0 migrations have already been run
+	// We know if 0.9.0 migrations have been run because dbCodeVersion will be populated if so
+	// TODO: related epic https://github.com/spiffe/spire/issues/1083
+	// This check can be removed in 0.10.0
+	if migration.CodeVersion == "" {
+		if disableMigration {
+			err := sqlError.New("must upgrade from version %s or higher to disable auto-migration", minimumCodeVersionToDisableMigrate)
+			log.Error(err.Error())
+			return err
+		}
+		// set to 0.0.0
+		dbCodeVersion = semver.Version{}
+	} else {
+		dbCodeVersion, err = semver.Parse(migration.CodeVersion)
+		if err != nil {
+			return err
+		}
 	}
 
-	if version == codeVersion {
+	if schemaVersion == latestSchemaVersion {
+		log.Debug(fmt.Sprintf("Code and DB schema versions are both %d. No migration needed.", schemaVersion))
+
+		// same DB schema; if current code version greater than stored, store newer code version
+		if codeVersion.GT(dbCodeVersion) {
+			newMigration := Migration{
+				Version:     latestSchemaVersion,
+				CodeVersion: codeVersion.String(),
+			}
+
+			if err := db.Model(&Migration{}).Updates(newMigration).Error; err != nil {
+				return sqlError.Wrap(err)
+			}
+		}
 		return nil
 	}
 
+	// If auto-migrate is disabled and we are running a compatible version (+/- 1
+	// minor from the stored code version) then we are done here
+	if disableMigration {
+		if !isCompatibleCodeVersion(dbCodeVersion) {
+			return sqlError.New("auto-migrate is disabled but current code version %s is not compatible with the current DB schema version %d",
+				codeVersion.String(), schemaVersion)
+		}
+		return nil
+	}
+
+	// The DB schema version can get ahead of us if the cluster is in the middle of
+	// an upgrade. So long as the version is compatible, log a warning and continue.
+	// Otherwise, we should bail out. Migration rollbacks are not supported.
+	if schemaVersion > latestSchemaVersion {
+		if !isCompatibleCodeVersion(dbCodeVersion) {
+			return sqlError.New("incompatible DB schema %d is too new for code version %s, upgrade SPIRE Server",
+				schemaVersion, codeVersion.String())
+		}
+		log.Warn(fmt.Sprintf("DB schema %d is ahead of code version %s with schema version %d, upgrading SPIRE Server is recommended",
+			schemaVersion, codeVersion.String(), latestSchemaVersion))
+		return nil
+	}
+
+	// at this point:
+	// - auto-migration is enabled
+	// - schema version of DB is behind
+	// - code version difference is not too large to support
```

Comment:
I think this comment line can be removed - code version could be 0 (if it is a fresh cluster). And while it will probably not happen often in practice, we could _technically_ still migrate you up from e.g. code version 0.9 to 0.42 in a safe fashion if we have reached this point.